### ISSUE TITLE
API support for dealing with interned strings directly

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1300]
When C code operates on object properties a C string is passed in (usually a literal but can also be a dynamic input). All such operations will be slowed down by the need to intern the string.

Internally property operations are ultimately always performed using interned strings (`duk_hstring *`) and since it's already possible to get a borrowed reference to a heap object, it would be straightforward to allow something like:

```c
duk_get_prop_interned(duk_context *ctx, duk_idx_t obj_idx, void *borrowed_string_ref);
```

The borrowed string ref would be obtained using `duk_get_heapptr()` and user code would be responsible for ensuring that the string remains reachable.

This would speed up property operations in critical paths quite significantly.

Naturally there are many API variations, like:
- Instead of using the generic `duk_get_heapptr()` provide a dedicated API call to intern a string.
- For long lived strings it might be useful to be able to `void *mystr = duk_something(ctx, "my_key");` which would store a stash (or other persistent) reference so that you wouldn't need to worry about string reachability.

But as an initial step, being able to use a borrowed key reference would be easiest.

**Comment by svaarala** [comment]
The descriptiond describes using borrowed heap references as the way to work with interned strings, but at a higher level there are many other options which would be viable.

Another approach would be safer string handles: one API call would intern a string and return a handle (a running number) and that handle would be used by the API calls. Even if the underlying string was freed for some reason, the handle would be memory safe and an attempt to use it would fail with an error.

The downside here being, of course, that direct heap references are faster than maintaining some sort of handle index.

**Comment by harold-b** [comment]
This is a nice feature :)

Speaking of performance, I tested out [scriptorium](https://github.com/r-lyeh/scriptorium) last night with duktape v2.0.0, compiled (MSVC) w/ `CL /c /MT /MP /GL /O2` and `LINK /LTCG /MACHINE:X86`, and the following duktape config:
```
-DDUK_USE_FASTINT
-UDUK_USE_VOLUNTARY_GC
-DDUK_USE_EXEC_FUN_LOCAL
-DDUK_USE_VALSTACK_UNSAFE
```
I made fib.js compile as function code and grabbed the print function (simply native binding) into a local var. The results were around **25 pt** with those options, disabling `DDUK_USE_EXEC_FUN_LOCAL` lowered it a bit. This was only tested with a 32bit runtime.

Nice boost! :+1:  This is compared to the version they were using, [which as you can see here, is **v1.2.2**](https://github.com/r-lyeh/scriptorium/blob/master/javascript/duktape/duktape.h#L8).

Do you think in the future we'd get to the Angelscript/Squirrel region? (I ask naively not knowing the requirements of the standard or the runtime that might *not* let that become a thing :) ).

**Comment by fatcerberus** [comment]
The main problem with optimizing JS engines is that ECMAScript is by nature very dynamic - even built-ins that are well-defined by the standard can change out from under you at any time due to polyfills, monkey patching, etc..  While this is generally considered a strength of the platform, it's difficult for an engine to make any assumptions when compiling the code.  So I would be very surprised if Duktape ever managed to match the performance of, say, Lua.

I could be wrong about that though, @svaarala has made some really big strides with performance recently. :)

**Comment by harold-b** [comment]
> JS engines is that ECMAScript is by nature very dynamic... ... So I would be very surprised if Duktape ever managed to match the performance of, say, Lua.

Yes, I realize that -- That's why I settled for Angelscript/Squirrel territory :grinning:. And put on the disclaimer as well since I know Saami has mentioned that the focus is small footprint over performance. But we can dream :)!

**Comment by svaarala** [comment]
> JS engines is that ECMAScript is by nature very dynamic... ... So I would be very surprised if Duktape ever managed to match the performance of, say, Lua.

@fatcerberus Plain Lua (not luajit) is also dynamically typed so there's not a huge structural difference in that respect. Duktape's bytecode dispatch is actually now relatively close to Lua's, but there are some areas that still have a lot of potential for optimization - like property handling and slow path variable handling (scopes in general). Lua semantics are also in general a bit more friendly for fast interpretation, and refcounting also has some cost.

The plan is to keep chipping away on these one by one. The reason there's potential for optimization is that I thought it more important to achieve a functional 1.0.0 rather than keep on rewriting the building blocks over and over for performance (which is also important, but having a baseline release is maybe more important).

Longer term many things like some limited form of type inference (say loop counters and other "local" cases) may be feasible. And while no JIT work is currently in progress, it's certainly possible at some point, provided that it doesn't conflict with portability and other major goals. Reworking the compiler for ES2015 also opens up many small avenues for optimization such as better constant folding, dead code elimination, etc.

So I look at it more as a continuous process than achieving a certain level and then be done with it :-)

**Comment by harold-b** [comment]
> Duktape's bytecode dispatch is actually now relatively close to Lua's

Would that mean a major point of optimization could come from the compiler's bytecode generation?

**Comment by svaarala** [comment]
> Would that mean a major point of optimization could come from the compiler's bytecode generation?

Not really, while the compiler output is not perfect it's generally quite OK. It's really property accesses, slow path variable accesses, call handling, etc that are slower.

For example, for an empty loop:

```
duktape@keppihevonen:~/duktape$ time ./duk.O2.pgo tests/perf/test-empty-loop.js 
real	0m1.374s
user	0m1.373s
sys	0m0.000s

duktape@keppihevonen:~/duktape$ time lua tests/perf/test-empty-loop.lua 
real	0m1.020s
user	0m1.014s
sys	0m0.000s
```

Lua is a bit faster, but for a straight loop just reference counting alone can account much of the difference; here's the same without refcounting:

```
duktape@keppihevonen:~/duktape$ time ./duk-perf-pgo.O2 tests/perf/test-empty-loop.js 

real	0m1.193s
user	0m1.192s
sys	0m0.000s
```

So it's not quite there (and to be fair, this against whatever Lua-build is in my distributable which is probably not PGO and so on), but achieving a dispatch parity is still probably achievable.

**Comment by harold-b** [comment]
> Lua is a bit faster, but for a straight loop just reference counting alone can account much of the difference;

Oh wow, very interesting. Thanks for the info.

What are the negative effects or trade offs of turning off `DUK_USE_REFERENCE_COUNTING` ?

**Comment by svaarala** [comment]
Memory usage will be more "sawtooth" like (rising, then dropping, rising again, etc) and finalizer execution will be postponed until mark-and-sweep. If you're not using a pool allocator and a dedicated pool, the rest of the system (sharing malloc() for example) can't exert memory pressure on Duktape even if an allocation fails and garbage was freeable. That issue does exist with reference counting but to a much lesser extent, affecting garbage which resides in reference loops.

**Comment by harold-b** [comment]
Sounds like a good fit for my game framework since the script system memory is specifically managed already :+1: 

I just tried out scriptorium with no ref counting, it went up by 2 points ( **27 pt** ). Nice!

**Comment by svaarala** [comment]
It'd be interesting to hear how things would work from master (if I understood correctly you're running against 2.0.0), as there are some improvements to the string table and the object hash table in master (and some more to come for 2.1.0 :-)

**Comment by harold-b** [comment]
Just tried a few times it with:
```
 *  Git commit a446bd8ce9949b29922e057f16e7813faef3e357 (v2.0.0-96-ga446bd8).
 *  Git branch master.
```

Got **26-28 pts**, but mostly 26. I only tested once earlier with no ref-count enabled to get the 27, so that one wasn't an average.

**Comment by svaarala** [comment]
Ok, so no difference for that test. The changes so far really only matter in string heavy code (code creating or manipulating a lot of strings) or object heavy code (code dealing with more than just very small objects).

**Comment by harold-b** [comment]
Got it. Will be interesting to see how you get duktape to climb up the ladder as continue adding optimizations! :)

**Comment by svaarala** [comment]
Hmm, just quickly looking up, the benchmark (scriptorium) seems limited to just Fibonacci which exercises call handling almost exclusively. There's plenty to improve there - but it's not very representative of practical performance.

**Comment by harold-b** [comment]
Oh I agree. Just thought it's a nice case study seeing as it goes across so many scripting engines.

At some point, when I'm not so time-bound to a project ( :disappointed:  ) I'd like to be able to add some profiling to to my framework so perhaps you can have some more real-world data at your disposal (for this type of app at least). Though I see that you recently added a benchmark issue with @fatcerberus's engine :)

**Comment by harold-b** [comment]
Having said that I've not yet had issues with any of my test Apps even on debug builds on my oldest device (iPhone 4S). They've all run at 60 FPS. Granted, these were small games have not been too complex on the scirpting side.

I did ran into slowdowns when making too many native calls in a loop. It was a case that could probably easily be optimized on my side by adding an API to use a TypedArray instead and pass in that whole value only once, but nonetheless this was the only case where I had some slowdown.

**Comment by svaarala** [comment]
Yeah, having some automated regression testing for typical application loads would be very useful. I'm mostly using microbenchmarks for commit testing and they're quite useful to detect changes in individual mechanisms. But they are not good indicators of the overall performance, and it's hard to find good benchmarks that actually match application behavior well. So any application benchmarks are quite welcome :-) They don't need to be multi-engine for Duktape commit measurement purposes - but it's of course useful if comparison to other engines/languages is possible.

**Comment by fatcerberus** [comment]
> I thought it more important to achieve a functional 1.0.0 rather than keep on rewriting the building blocks over and over for performance (which is also important, but having a baseline release is maybe more important).

Oh, I agree 100%.  If it weren't for Duktape, Spherical *still* wouldn't have a modern engine, so I'm glad you released it :)

**Comment by svaarala** [comment]
The downside is more frequent mark-and-sweeps and a cluster of free()s in each, as opposed to the steadier free()s of refcounting. Refcounting promotes cache locality to some degree because the freed memory slots are often reused. But the net effect is pretty complicated to estimate, so ultimately you would need to test it in practice.

Ideally Duktape's mark-and-sweep would be incremental; it would then be a much better option for your situation. I ran into some trouble originally with incremental GC and refcounting, so I opted for the simpler non-incremental mark-and-sweep to first get things working. It might be a good time to revisit that. Lot of stuff to do besides that, though :-)

**Comment by harold-b** [comment]
While we're on the subject of bombarding @svaarala with performance queries... Here's one more...

Is it slower yo do property access of a getter function than a standard property?
To be more specific, in my engine, small objects that get re-used a lot are pooled. I pooled some math objects (Vectors, Quaternions) for the lifetime of the duktape context, so that they can be re-used continually. However, right now, my math objects are just a native binding, including their properties (A Vector2 has getters defined for x,y and an index into to a native Vector struct from the pool). Is accessing their properties with a getter slower than if I'd simply defined the x/y as regular object properties in the Ecmascript object itself? Then of course there would be the added overhead to native functions (like getting a Vector's length, for example) which would have to lookup the x/y properties cast to 32-bit float and then perform operations on it. 

But in the case that most of their use is reading/writing the x/y properties of the vectors from Ecmascript, which path is the better one regarding performance?

**Comment by svaarala** [comment]
Getters would be slower because they involve an actual function call which needs to reconfigure the value stack etc.

**Comment by harold-b** [comment]
Doh! Thought that would be the case. Will have to re-write that.

Thanks Sami.

**Comment by svaarala** [comment]
There's currently no convenient way to have a mutable native vector which is also efficiently accessible from the script side: either the vector is native and getters or native calls are needed to access the fields, or the vector exists in the script side, and native access needs property reads.

The concept missing from this situation would be a form of very lightweight Proxy where the hooks were native functions that don't involve a call stack visible function call. Instead, an integer index read, for example, would invoke a native call directly and it would just push a number to the value stack and return.

This is something that I'd like to use internally too, for implementing some of the Ecmascript magic behaviors, instead of the current monolithic property code.

**Comment by svaarala** [comment]
Well, what I wrote above is true for the general native/script object mapping use cases.

But in your specific use cases, maybe you could use Float64Array to represent the vectors. That can be backed to a buffer which is also visible 1:1 to native code so both native code and scripts can read/write the indexed properties with a minimum of fuss.

**Comment by harold-b** [comment]
> But in your specific use cases, maybe you could use Float64Array to represent the vectors

I was just about to ask about this, haha.
The previous comment would certainly useful, if it gets implemented at some point.

Regarding the TypedArray -- The problem with this is that one would have to use [0]/[1] in place of .x/.y, right? Which would not be very practical from the 'programmer's (me in this case) point of view, as it's not an intuitive syntax to use with Vectors in general.

**Comment by svaarala** [comment]
Yep, you would need to use integer indices.

**Comment by harold-b** [comment]
Out of curiosity (not a hint to beg for the feature, just curios), how could the API you mentioned above be implemented? Could it piggyback the buffer API behind the scenes in a way and have some sort of virtual property access into the buffer's index?

**Comment by harold-b** [comment]
For example, could a 'struct' object, if you will, be an object backed by a buffer which could be accessed instead by virtual properties that are mapped to indices (alignment issues notwithstanding).

**Comment by svaarala** [comment]
If there was a lightweight Proxy-like hook API, the basic pattern might be:
- Construct a lightweight Proxy object which would be just an empty object (initially) but with a table for native callbacks/hooks for property read, write, delete, etc.
- When a property read (for example) occurs, property code would detect it's dealing with a lightweight proxy and call the native callback. Value stack top would provide the key, and the callback would replace the key with the property value.

If there was an ability for the native code to intern strings, it would be able to check the string by comparing string pointers:

```c
void *p = duk_get_heapptr(ctx, -1);
if (p == my_interned_x) {
} else if (p == my_interned_y) {
}
/* ... */
```

Alternatively, the C function could get the interned string `void *` as a direct argument (avoiding the `duk_get_heapptr()`):

```c
if (arg_key == my_interned_x) {
    /* ... */
}
```

The callback could also be given a `const char *` + len instead which is more native to C code.

A few more variations on this theme:
- Property read code would first consult the property table on the lightweight Proxy. If the property is present, it would be read from there. This would allow virtualization of properties with the property table acting as a cache (makes sense if the property value is expensive to calculate).
- Instead of hooks getting a key argument, one might register all the virtual properties into the property table, with the property value being a plain pointer to a native callback table for that property. This would allow, for example, a separate callback to be used for property `.x`, `.y`, and `.z`. This would work much better for a large amount of properties - though if a bunch of objects is created with the same virtual property callbacks, these would ideally be reused somehow. For a large number of properties there would need to be a hash table to quickly locate the relevant callbacks.
- Yet another variation would be to use the lightweight Proxy property table to store a string -> integer mapping, for example map x -> 0, y -> 1, z -> 2. Property lookup code would look up this integer and give it to the native callback. In your case, the indices would map to coordinates and be easy to use in a switch-case. Further, the property table already provides a hash map so that if there are a lot of mapped variables, they can be efficiently looked up (and inherited from an ancestor if necessary).

So, there are quite many variations for this - and it would be nice to find one variation that would work for a lot of use cases so that multiple similar but ultimately different mechanisms could be avoided.

**Comment by svaarala** [comment]
> For example, could a 'struct' object, if you will, be an object backed by a buffer which could be accessed instead by virtual properties that are mapped to indices (alignment issues notwithstanding).

This would be possible for a limited set of data types, but it would need a bit of special handling to deal with strings for example. Struct <-> object mapping is a common use case though so maybe something like this would be useful even with some limitations.

But at least as a first step I'd be more looking into generic property access virtualization, and making that as efficient and C-native as possible.

**Comment by harold-b** [comment]
This is all very interesting! I wasn't even thinking of this type of method, this would certainly be useful. How does this effectively differ to say, a regular native function? (the first method you demonstrated). I know you  always only need 1 value in the stack, so there's no need to prepare for variation there. But which other checks are bypassed which make it faster than a regular native call?

What I had in mind was more of a struct mapping API, where the only virtualization was to get an index into a buffer. So that structs could be transparently (sort of) shared between native and Ecmascript.

For example, the aforementioned Vector2 could be mapped in something like this:

``` C
duk_push_struct_object(ctx, sizeof(Vector2)); /* [struct] */
/* Or the user can even specify an external buffer with a size */

duk_put_struct_prop_string(ctx, -1, "x", DUK_SPROP_FLOAT32, offsetof(Vector2, x));
duk_put_struct_prop_string(ctx, -1, "y", DUK_SPROP_FLOAT32, offsetof(Vector2, y));

/* union access */
duk_put_struct_prop_index(ctx, -1, 0, DUK_SPROP_FLOAT32, offsetof(Vector2, x));
duk_put_struct_prop_index(ctx, -1, 1, DUK_SPROP_FLOAT32, offsetof(Vector2, y));
```

Then on property access from the Ecmascript side it could look up the offset and type and coerce it into a number, no need to call a user function. From the C side, the user can just use `Vector2* v = (Vector2*)duk_require_buffer_data()` and modify it.

I don't know the duktape internals well enough so I don't even know how feasible that is, but certainly would be nice, and it seems the Buffer system has some of that stuff in place already?

**Comment by svaarala** [comment]
Call handling involves a lot of small steps like:
- Size checking and potentially resizing the call stack
- Dealing with the "this" binding
- Dealing with "arguments"
- Resizing the value stack and changing the value stack "bottom" offset
- Managing expected vs. actual argument and return value counts
- Resizing the value stack back to the caller's state, and restoring the value stack "bottom" offset

Duktape internals do a lot of calls where the helper consumes a few values from the value stack and push result value(s) in return. So this kind of hook would be no less efficient than any of these internal calls, except for needing to use the standard ABI and not being able to be inlined.

**Comment by harold-b** [comment]
Very informative, so that would certainly be a nice boost. :)

I think applications like mine, which do a large portion of their work by calling native bindings, including many getters/setters, would benefit well across the board with that API. I'll have to keep an eye out for that one as well.

**Comment by svaarala** [comment]
Re: creating a "struct object", it would be conceptually clear to do that.

Two basic questions there would be:
- Where to store the reference to the underlying buffer? It might require a `duk_hobject` "subclass" similar to `duk_hbufobj` which is an extension of the generic object structure. So, solvable.
- The field/offset + metadata info would need to be stored somewhere. The standard property table is just key/value + 8-bit attributes. The metadata could maybe be packed into a single integer, e.g. as 8-bit type/flags and 24-bit offset (up to 53 bits could be used when using IEEE doubles). If that's not sufficient, a different data structure would be needed.

So all in all this would be doable, but such a mechanism might not be generic enough to be worth its footprint: if at any point you need to virtualize any of the accesses (write protect, process/validate/log values read, etc) some form of callbacks are needed in any case.

Just to compare, one lightweight Proxy variant would use the property table to map a property name to an arbitrary integer passed to the C callback. For example, `.x` could map to 0, `.y` to 1, etc. Then, the native call would look something like:

```c
/* Value stack behavior: no value stack input, push one value stack output. */
void my_read_hook(duk_context *ctx, duk_int_t int_arg) {
    switch (int_arg) {
    case 0:
        duk_push_number(ctx, ...);  // push X
        break;
    case 1:
        duk_push_number(ctx, ...);  // push Y
        break;
    default:
        duk_push_number(ctx, ...);  // push Z
    }
}
```

But since the integer was arbitrary, it could just as easily encode e.g. a byte width and and offset, e.g. as `(width << 24) + offset`, and the callback could then:

```c
void my_read_hook(duk_context *ctx, duk_int_t int_arg) {
    int width = int_arg >> 24;
    int offset = int_arg & 0xffffffL;

    /* Read 'width' bytes from 'offset'. */

    /* XXX: not sure how the underlying buffer reference would be obtained here.
     * Maybe the value stack should contain the object; or the call arguments could
     * provide a value stack index for the object.
     */
}
```

The nice thing here would be that it would be up to the application to populate the property name -> integer entries as ordinary properties, and then use the integers in whatever way worked well for the native call.

But as said above, there are a lot of different use cases and variations, and figuring out a reasonably generic mechanism (or set of mechanisms) that's usable in a variety of scenarios is a real challenge :)

**Comment by svaarala** [comment]
> I think applications like mine, which do a large portion of their work by calling native bindings, including many getters/setters, would benefit well across the board with that API. I'll have to keep an eye out for that one as well.

Certainly, so far the approach has been to direct effort to Proxy and other standard features. But ultimately some object mapping scenarios need some more direct native integration. But what I do want to avoid is adding a lot of similar mechanisms because each involves footprint, documentation, etc, so an important question is to ensure a minimal set of very widely applicable mechanisms are added.

**Comment by svaarala** [comment]
@harold-b We should continue this discussion in #1302, I opened that for tracking the native virtualization use case separately.

**Comment by svaarala** [comment]
Regarding the issue in the title, it's already possible to use interned strings for property accesses via the value stack. For example:

Setup phase:
```c
/* Setup. */
void *my_key;
duk_push_global_stash(ctx);
duk_push_string(ctx, "foo");
my_key = duk_get_heapptr(ctx, -1);
duk_put_prop_string(ctx, 2, "foo");  /* stash['foo'] = 'foo', just to make the string referenced, could be stash['foo'] = 1 etc. */
```

And then to e.g. read a property:

```c
duk_push_heapptr(ctx, my_key);  /* 'foo' */
duk_get_prop(ctx, obj_idx);
```

The improvement in this issue would be something like:

```c
duk_get_prop_heapptr(ctx, -3, my_key);
```

(Naming up for grabs of course, e.g. `duk_get_prop_interned()` makes more sense if there is a corresponding API key to explicitly intern a string.)

**Comment by svaarala** [comment]
(Updated description to describe the currently possible interned string usage.)

**Comment by harold-b** [comment]
> Regarding the issue in the title, it's already possible to use interned strings for property accesses via the value stack. 

This is good to know, I wasn't aware the heapptr for the string represented the interned value.

**Comment by svaarala** [comment]
Right, in the current typing model a heap pointer can point to a string (`duk_hstring`), an object (`duk_hobject` and its various subclasses), or a plain buffer (`duk_hbuffer` and its subclasses). This is not very prominent in the public documentation because it's more of an internal detail.

**Comment by harold-b** [comment]
And this work with internal properties as well, right?

**Comment by svaarala** [comment]
Yes, there's no difference from the internals point of view.

**Comment by tlsa** [comment]
> (Naming up for grabs of course, e.g. `duk_get_prop_interned()` makes more sense if there is a corresponding API key to explicitly intern a string.)

I think that would be better.  The setup phase in your example code isn't as clear as e.g. `duk_create_interned()`.
